pandas/core/indexes/base.py

```diff
@@ -3890,6 +3890,7 @@ def _evaluate_compare(self, other):
                 except TypeError:
                     return result
 
+            _evaluate_compare.__name__ = '__{name}__'.format(name=op.__name__)
```

Comment:
we have a function in core/generic for exactly this

(and sets properly in py2/3 and such)

Reply:
Closest I can think of is pd.compat.bind_method.  Is that what you're referring to?  Wrapped method names are set like this in e.g. indexes.datetimes._field_accessor.

Reply:
use pandas.compat.set_function_name

Reply:
That's going to get into a circularity problem.  set_function_name requires `cls`, but in several cases this is being called before the class has been defined in the namespace.

Scroll up to line 50 in timedeltas for an example of the status quo usage.

Reply:
see how its done in generic.py, instead you should change these to a factor function.

Reply:
OK I'll do it, but under protest... those classmethods to _add_foo_methods are among my least favorite design patterns.

Reply:
these are already a well established pattern

and they handle the edge cases

eg qualname

Reply:
In indexes.base.Index the `func.__name__ = name` pattern is used in `_add_logical_methods`, `_add_logical_methods_disabled`, `_add_numeric_methods_disabled`, `_add_numeric_methods_add_sub_disabled`.  In case you want to open an issue to change those.